adding ps_strategy to run_config to enable different placement strate…

dave-msk opened the pull request:
…gy in estimator

This is my first pull request. I noticed that there isn't any way to set the `ps_strategy` in `tf.train.replica_device_setter` if `tf.estimator.Estimator` is used in distributed settings. I guess the best way to notify the estimator which strategy to use is through the `RunConfig` class. I have added a property `ps_strategy` to `RunConfig` and filled in some of the unit tests. I have also modified the `_get_replica_device_setter` in `estimator.py` to return a device setter with the `ps_strategy` included. Please kindly review and let me know if there is anything I need to correct or modify

Commit:
adding ps_strategy to run_config to enable different placement strategy in estimator

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented:
Jenkins, test this please.

martinwicke commented:
The way to achieve this is the device_fn. There's already a private field in Estimator (_device_fn) which is used for that purpose. 

As a hack you can overwrite this after construction, but if we want to make this official, I would suggest we add a @property device_fn() to RunConfig, which is used to create the _device_fn field instead of the module private function used at the moment. You could then change what that property returns to change the PS placement strategy.

@xiejw WDYT?

xiejw commented:
Thanks @dave-msk.

@martinwicke  This is fine. I prefer device_fn over ps_strategy due to the fact that device_fn is more general but ps_strategy assumes ps is not the cluster.

However, it will be user's own risk to debug this. device_fn or ps_strategy is not easy to write.

dave-msk commented:
@martinwicke @xiejw ,
Thanks for reviewing. It makes sense to me to add @property to Run_Config() that client code could just do this by an extra `config.device_fn = foo`. I'll try to get this updated asap

Commit:
Resolved merge conflict on import lines

dave-msk commented:
I see the current approach for distributed training is that the RunConfig first parses the `TF_CONFIG` environment variable to get details like `master`, `task_id`, `num_ps_replicas` etc. These are then used to create a replica device setter when Estimator is constructed through the `training.replica_device_setter` call. To allow a dynamic `device_fn` override while preserving the convenience of the current approach, I'd suggest we do it this way:

1. We add a @property `device_fn` to RunConfig, set it `None` as default
2. We call `self._config.device_fn` to get the most updated version of `device_fn` in `_train_model` in Estimator.
3. If the `device_fn` is `None` (not overridden by user) in RunConfig, we use the current approach (via `training.replica_device_setter` call) to return a "default" device function, with the `ps_strategy` I added in my previous commit.
4. If the `device_fn` is not `None` (overridden by user) in RunConfig, we return it directly.

If one chooses to write his own device function, he would have to define the `device_fn` (`config.device = foo`) and take care of everything manually. If the user just wanted to specify a `ps_strategy` other than the default round-robin (e.g. the `tf.contrib.training.GreedyLoadBalancingStrategy`), he could just construct the RunConfig with `ps_strategy` or do a replace (`config = config.replace(ps_strategy=ps_strategy)`).

@xiejw @martinwicke what do you think?

Commit:
1. Moved estimator._device_fn to RunConfig as @property
2. Made RunConfig.device_fn to return custom device function if one is specified, otherwise the result from `tf.train.replica_device_setter` call is used
3. Added some basic unit tests, may need further tests.

xiejw commented on tensorflow/python/estimator/estimator.py:
```diff
@@ -207,7 +207,7 @@ def __init__(self, model_fn, model_dir=None, config=None, params=None,
       self._session_config = self._config.session_config
 
-    self._device_fn = _get_replica_device_setter(self._config)
+    # self._device_fn = _get_replica_device_setter(self._config)
```
I think we might need to keep this line such that 
_get_replica_device_setter should respect config.device_fn first and, if missing, create a good default like the old behavior. 

wdyt?

xiejw commented on tensorflow/python/estimator/estimator.py:
```diff
@@ -811,7 +811,7 @@ def _call_model_fn(self, features, labels, mode, config):
   def _train_model(self, input_fn, hooks, saving_listeners):
     worker_hooks = []
-    with ops.Graph().as_default() as g, g.device(self._device_fn):
+    with ops.Graph().as_default() as g, g.device(self._config.device_fn):  # g.device(self._device_fn):
```
remove the comment in the tail?

xiejw commented on tensorflow/python/estimator/estimator.py:
```diff
@@ -1025,7 +1025,8 @@ def _get_replica_device_setter(config):
         ps_ops=ps_ops,
-        cluster=config.cluster_spec)
+        cluster=config.cluster_spec,
+        ps_strategy=config.ps_strategy)
```
you might need to adjust here. I do not feel we need to have ps_strategy here.

xiejw commented on tensorflow/python/estimator/run_config.py:
```diff
@@ -43,7 +47,9 @@
-    'log_step_count_steps'
+    'log_step_count_steps',
+    'device_fn',
+    'ps_strategy'
```
drop ps_strategy?

xiejw commented on tensorflow/python/estimator/run_config.py:
```diff
@@ -27,11 +27,15 @@
 
 
 _USE_DEFAULT = object()
+_VALID_PS_STRATEGY_ARGS = set(['op'])
```
I did not see these two were using in the file. Did I miss anything?

xiejw commented on tensorflow/python/estimator/run_config.py:
```diff
@@ -582,6 +622,10 @@ def num_ps_replicas(self):
     return self._num_worker_replicas
 
+  @property
+  def ps_strategy(self):
```
drop this?

xiejw commented:
Hi @dave-msk I think most of part of your suggestions sound great. 

We should allow user to specify the device_fn in RunConfig and respect that. If the device_fn is None, we create a default one (old behavior). 

I left some comments in the files. I think we should drop the ps_strategy in the RunConfig as device_fn provides higher level support for that. Advanced users could create device_fn directly and use different ps_strategy.

Commit:
1. Removing ps_strategy.
2. Modified estimator to take overriden device_fn from  if set.
3. Removed ps_strategy related unit tests.

dave-msk commented:
The `ps_strategy` is removed. So the usage will now be as follows:

1. The user has a self-defined device function `device_fn`:
```
config = RunConfig(device_fn=device_fn)
```

2. The user wants to change only the ps strategy to `ps_strategy` in the default device function used by `Estimator`. In this case the easiest way is to use the code from `_get_replica_device_setter(config)` function in `estimator.py`, with an addition parameter `ps_strategy`:

```
def get_replica_device_setter(config, ps_strategy):
  ps_ops = [
      'Variable', 'VariableV2', 'AutoReloadVariable', 'MutableHashTable',
      'MutableHashTableV2', 'MutableHashTableOfTensors',
      'MutableHashTableOfTensorsV2', 'MutableDenseHashTable',
      'MutableDenseHashTableV2', 'VarHandleOp'
  ]

  if config.task_type:
    worker_device = '/job:%s/task:%d' % (config.task_type, config.task_id)
  else:
    worker_device = '/job:worker'

  if config.num_ps_replicas > 0:
    return training.replica_device_setter(
        ps_tasks=config.num_ps_replicas,
        worker_device=worker_device,
        merge_devices=True,
        ps_ops=ps_ops,
        cluster=config.cluster_spec,
        ps_strategy=ps_strategy)  # This is the only difference from the original code
  else:
    return None

config = RunConfig()
config = config.replace(device_fn=get_replica_device_setter(config, ps_strategy))
```

3. If the user wants to use the default device function, no code change is required.

Adding a `device_fn` gives a great flexibility to users, but it doesn't feel right if we need to dive deep into the source code and define a slightly modified version of `_get_replica_device_setter` if the ps_strategy is the only thing one wants to change. What do you think?

xiejw commented:
Thanks. The code looks good to me. This is a tradeoff we need to decide. Currently, Estimator mostly works with ps in distributed model. But in future, this assumption might not hold anymore. Device_fn provides a better abstraction for that. 

Regarding the issue changing ps_strategy: The get_replica_device_setter can be much simplified as

 def get_replica_device_setter(config, ps_strategy):
  ps_ops = [
      'Variable', 'VariableV2', 'AutoReloadVariable', 'MutableHashTable',
      'MutableHashTableV2', 'MutableHashTableOfTensors',
      'MutableHashTableOfTensorsV2', 'MutableDenseHashTable',
      'MutableDenseHashTableV2', 'VarHandleOp'
  ]

   worker_device = '/job:%s/task:%d' % (config.task_type, config.task_id) 
   return training.replica_device_setter(
        ps_tasks=config.num_ps_replicas,
        worker_device=worker_device,
        merge_devices=True,
        ps_ops=ps_ops,
        cluster=config.cluster_spec,
        ps_strategy=ps_strategy)  # This is the only difference from the original code
 

As user is changing the ps_strategy, all the if-else branches are not necessary. The only remaining silly thing here is the ps_ops list. I think it should not be defined here and user needs to copy/paste. I will move it as a constant list to replica_device_setter. such that user can easily refer them. With that, user code could be significantly reduced. Does this sound reasonable?

dave-msk commented:
Makes sense. So I guess the code becomes:

```
def get_replica_device_setter(config, ps_strategy):
  worker_device = '/job:%s/task:%d' % (config.task_type, config.task_id) 
  return training.replica_device_setter(
      ps_tasks=config.num_ps_replicas,
      worker_device=worker_device,
      merge_devices=True,
      ps_ops=tf.train.some_ref_to_ps_ops,  # just some reference to the original ps_ops
      cluster=config.cluster_spec,
      ps_strategy=ps_strategy)
```

It looks cool, couldn't think of a better way to do it right now

xiejw reviewed (approved):
Thanks for the contribution and multiple rounds of iterations!

josh11b commented:
This is good for API review

dave-msk commented:
I have noticed that the `device_fn` is implemented using the variable `distribute`. Should I change the name to resolve the possible merge conflict? @xiejw

Commit:
Resolving merge conflict

drpngx commented:
Need to update the API goldens:

```

    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

drpngx commented:
Also errors such as this:

```
ERROR: testPoissonRegression (__main__.LinearEstimatorTest)
Tests that loss goes down with training.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/contrib/learn/python/learn/estimators/linear_test.py", line 1792, in testPoissonRegression
    head=head_lib.poisson_regression_head())
FAIL: //tensorflow/contrib/learn:debug_test (shard 3 of 4) (see /Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/testlogs/tensorflow/contrib/learn/debug_test/shard_3_of_4/test.log)
INFO: From Testing //tensorflow/contrib/learn:debug_test (shard 3 of 4):
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/contrib/learn/python/learn/estimators/linear.py", line 980, in __init__
    feature_engineering_fn=feature_engineering_fn)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1179, in __init__
    super(Estimator, self).__init__(model_dir=model_dir, config=config)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/python/util/deprecation.py", line 250, in new_func
    return func(*args, **kwargs)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 455, in __init__
    self._config = self._config.replace(model_dir=self._model_dir)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/python/estimator/run_config.py", line 744, in replace
    **kwargs)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/python/estimator/run_config.py", line 779, in _replace
    _validate_properties(config)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/python/estimator/run_config.py", line 288, in _validate_properties
    message='device_fn must be callable with exactly one argument "op".')
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/python/estimator/run_config.py", line 257, in _validate
    property_value = getattr(run_config, property_name)
  File "/Volumes/BuildData/tmpfs/bazel_output/_bazel_kbuilder/d1b2600cd78e76a92812a06683f5de10/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/contrib/learn/linear_test.runfiles/org_tensorflow/tensorflow/python/estimator/run_config.py", line 593, in device_fn
    return self._device_fn
AttributeError: 'RunConfig' object has no attribute '_device_fn'
```

Commit:
Adding manual initialization of _device_fn in legacy RunConfig class

dave-msk commented:
Added manual initialization to old RunConfig, please review. @xiejw

xiejw commented:
LGTM. Thanks for the fix. I have kicked the tests again for you.

Commit:
Merge branch 'master' of https://github.com/tensorflow/tensorflow

Commit:
Updated estimator golden API through
1. bazel build //tensorflow/tools/api/tests:api_compatibility_test
2. bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True

dave-msk commented:
The golden API of estimator is updated through
```
1. bazel build //tensorflow/tools/api/tests:api_compatibility_test
2. bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
```
This should pass the API test I think. Please review @xiejw

Commit:
Resolving merge conflict

Commit:
fixing code styles

Commit:
Merge branch 'master' of https://github.com/tensorflow/tensorflow

xiejw commented:
@jhseu please let me know if I need review here. I think the code was looking good but got a merge conflict later. I reviewed again. The current status LGTM
